### PR TITLE
PR: Stop automatic scrolling when a plot is selected (Plots)

### LINKED ISSUE #20474: Generating new plots always overrides currently selected plot
Consider the following code:

```
import matplotlib.pyplot as plt, time
for i in range(100):
    plt.figure()
    plt.plot([0, 1, i])
    plt.show()
    time.sleep(0.1)
```

Every time a new plot is generated, it is selected in the plots panel. If you want to look at a plot, you must wait untill all the plots are generated.

It would be great if, when you select a plot, the selected plot is not replaced by a newly plotted plot.

### PR DESCRIPTION
**Pull request by impact27**
## Description of Changes
Only scroll down in the plots plane if the latest plot is selected or if a new execution started

* [ ] Wrote at least one-line docstrings (for any new functions)
* [ ] Added unit test(s) covering the changes (if testable)

* [ ] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))

### Issue(s) Resolved

Fixes #20474

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct:

### CODE REVIEW

File: spyder/app/tests/test_mainwindow.py [hunk 1]
```diff
@@ -1991,6 +1991,77 @@ def test_plots_plugin(main_window, qtbot, tmpdir, mocker):
...
 
 
+def test_plots_scroll(main_window, qtbot):
+    """Test plots plugin scrolling"""
+    shell = main_window.ipyconsole.get_current_shellwidget()
+    figbrowser = main_window.plots.current_widget()
+
+    # Wait until the window is fully up.
+    qtbot.waitUntil(lambda: shell._prompt_html is not None,
+                    timeout=SHELL_TIMEOUT)
```
Comment (ccordoba12) [inline spyder/app/tests/test_mainwindow.py:2001]:
```suggestion
    qtbot.waitUntil(
        lambda: shell.spyder_kernel_ready and shell._prompt_html is not None,
        timeout=SHELL_TIMEOUT)
```

File: spyder/app/tests/test_mainwindow.py [hunk 2]
```diff
@@ -1991,6 +1991,77 @@ def test_plots_plugin(main_window, qtbot, tmpdir, mocker):
...
+    with qtbot.waitSignal(shell.executed):
+        shell.execute(("import matplotlib.pyplot as plt\n"
+                       "fig = plt.plot([1, 2, 3, 4], '.')\n"))
+
+    # Make sure the plot is selected
+    sb = figbrowser.thumbnails_sb
+    assert len(sb._thumbnails) == 1
+    assert sb._thumbnails[-1] == sb.current_thumbnail
+
+    # plot 4 more plots
```
Comment (ccordoba12) [inline spyder/app/tests/test_mainwindow.py:2013]:
```suggestion
    # Generate 4 more plots
```

File: spyder/app/tests/test_mainwindow.py [hunk 3]
```diff
@@ -1991,6 +1991,77 @@ def test_plots_plugin(main_window, qtbot, tmpdir, mocker):
...
+        shell.execute(
+            "for i in range(4):\n"
+            "    plt.figure()\n"
+            "    plt.plot([1, 2, 3, 4], '.')")
+
+    # we now have 5 plots and the last one is selected
+    assert len(sb._thumbnails) == 5
+    assert sb._thumbnails[-1] == sb.current_thumbnail
+
+    # plot 20 plots
```
Comment (ccordoba12) [inline spyder/app/tests/test_mainwindow.py:2024]:
```suggestion
    # Generate 20 plots
```

File: spyder/app/tests/test_mainwindow.py [hunk 4]
```diff
@@ -1991,6 +1991,77 @@ def test_plots_plugin(main_window, qtbot, tmpdir, mocker):
...
+            "    plt.figure()\n"
+            "    plt.plot([1, 2, 3, 4], '.')")
+
+    # Make sure we scrolled down
+    scrollbar = sb.scrollarea.verticalScrollBar()
+    assert len(sb._thumbnails) == 25
+    assert sb._thumbnails[-1] == sb.current_thumbnail
+    assert scrollbar.value() == scrollbar.maximum()
+
+    # plot 20 plots and select a plot in the middle
```
Comment (ccordoba12) [inline spyder/app/tests/test_mainwindow.py:2037]:
```suggestion
    # Generate 20 plots and select a plot in the middle
```

File: spyder/app/tests/test_mainwindow.py [hunk 5]
```diff
@@ -1991,6 +1991,77 @@ def test_plots_plugin(main_window, qtbot, tmpdir, mocker):
...
+            "for i in range(20):\n"
+            "    plt.figure()\n"
+            "    plt.plot([1, 2, 3, 4], '.')\n"
+            "    plt.show()\n"
+            "    time.sleep(.1)")
+        qtbot.wait(500)
+        sb.set_current_index(5)
+        scrollbar.setValue(scrollbar.minimum())
+
+    # make sure we didn't scroll to the end and a new thumnail was not selected
```
Comment (ccordoba12) [inline spyder/app/tests/test_mainwindow.py:2050]:
```suggestion
    # make sure we didn't scroll to the end and a new thumbnail was not selected
```

File: spyder/app/tests/test_mainwindow.py [hunk 6]
```diff
@@ -1991,6 +1991,77 @@ def test_plots_plugin(main_window, qtbot, tmpdir, mocker):
...
+        shell.execute(
+            "import time\n"
+            "for i in range(20):\n"
+            "    plt.figure()\n"
+            "    plt.plot([1, 2, 3, 4], '.')\n"
+            "    plt.show()\n"
+            "    time.sleep(.1)")
+        qtbot.wait(500)
+        sb.set_current_index(5)
+        scrollbar.setValue(scrollbar.minimum())
```
Comment (ccordoba12) [inline spyder/app/tests/test_mainwindow.py:2048]: I don't understand this part: shouldn't you call these two methods before generating 20 more plots and not here?

File: spyder/plugins/plots/widgets/figurebrowser.py [hunk 7]
```diff
@@ -240,6 +240,7 @@ def set_shellwidget(self, shellwidget):
         self.shellwidget = shellwidget
         shellwidget.set_figurebrowser(self)
         shellwidget.sig_new_inline_figure.connect(self._handle_new_figure)
+        shellwidget.executing.connect(self._handle_executing)
```
Comment (ccordoba12) [inline spyder/plugins/plots/widgets/figurebrowser.py:243]:
```suggestion
        shellwidget.executing.connect(self._handle_new_execution)
```

File: spyder/plugins/plots/widgets/figurebrowser.py [hunk 8]
```diff
@@ -248,6 +249,10 @@ def _handle_new_figure(self, fig, fmt):
         """
         self.thumbnails_sb.add_thumbnail(fig, fmt)
 
+    def _handle_executing(self):
+        """A new execution started"""
```
Comment (ccordoba12) [inline spyder/plugins/plots/widgets/figurebrowser.py:253]:
```suggestion
    def _handle_new_execution(self):
        """Handle a new execution in the console."""
```

File: spyder/plugins/plots/widgets/figurebrowser.py [hunk 9]
```diff
@@ -774,16 +789,27 @@ def add_thumbnail(self, fig, fmt):
...
         self.scene.setRowStretch(self.scene.rowCount() - 1, 0)
         self.scene.addWidget(thumbnail, self.scene.rowCount() - 1, 0)
         self.scene.setRowStretch(self.scene.rowCount(), 100)
-        self.set_current_thumbnail(thumbnail)
+        # Only select the new tumbnail if the last one was selected
+        select_last = (
+            len(self._thumbnails) < 2 or
+            self.current_thumbnail == self._thumbnails[-2] or
+            is_first
+        )
```
Comment (ccordoba12) [inline spyder/plugins/plots/widgets/figurebrowser.py:803]:
```suggestion

        # Only select a new thumbnail if the last one was selected
        select_last = (
            len(self._thumbnails) < 2
            or self.current_thumbnail == self._thumbnails[-2]
            or is_first
        )
```

**Review by ccordoba12 — COMMENTED** [review]
Thanks @impact27 for your work on this!

**Comment by ccordoba12** [comment]
Since this is not a critical bugfix, it needs to go against `master` (a few weeks ago we decided that everything non-critical should be left for Spyder 6).

For that, you need to rebase your branch with the following commands:

    git checkout stick_to_end_plots
    git rebase --onto master 5.x stick_to_end_plots
    git push -f origin stick_to_end_plots

File: spyder/app/tests/test_mainwindow.py (hunk 6 above)
Comment (impact27) [reply spyder/app/tests/test_mainwindow.py:2049]: I am calling these in the middle of generating 20 plots

Comment (impact27) [reply spyder/app/tests/test_mainwindow.py:2049]: I make 20 plots, one every 100ms. after 500ms I select a plot, so it should be during plotting

File: spyder/app/tests/test_mainwindow.py [hunk 10]
```diff
@@ -2102,7 +2103,7 @@ def test_plots_scroll(main_window, qtbot):
             "    plt.plot([1, 2, 3, 4], '.')\n"
             "    plt.show()\n"
             "    time.sleep(.1)")
-        qtbot.wait(500)
+        qtbot.waitUntil(lambda: sb._first_thumbnail_shown, timeout=SHELL_TIMEOUT)
```
Comment (ccordoba12) [inline spyder/app/tests/test_mainwindow.py:2106]:
```suggestion
        qtbot.waitUntil(lambda: sb._first_thumbnail_shown,
                        timeout=SHELL_TIMEOUT)
```

**Review by ccordoba12 — COMMENTED** [review]
@impact27, last minor suggestion for you, then this should be ready.

**Pull request merged**